### DISCUSSION TITLE
What makes wezterm bad?

### DISCUSSION THREAD

**Discussion by exastone** [wezterm/wezterm#2999] [General]
Hold up, before getting defensive I'm asking **"what make wezterm not the best terminal emulator?"**
I think wezterm might be the best terminal emulator currently and surprised it hasn't gained more attention, I was wondering what are the biggest limitations at the moment?

**Comment by jrmoulton** [comment]
I haven't found any limitations. It's great

**Reply by exastone** [reply]
It does seem pretty darn good, and i haven't come across anything besides the odd tweak here and there can't be addressed in the config file. I've seen some posts around attempting to benchmark performance of various terminal emulators (wezterm is usually not included in the comparison) and most of which provide the disclaimer that it's not a trivial task to provide an apples to apple comparison.

Alacrity has vtebench tool for benchmarking terminal emulator PTY read performance, I guess a better question would be has anyone had a go a benchmarking wezterm and found anything interesting?

**Comment by lifeym** [comment]
Thers's a benchmarking on Warp's official website: https://docs.warp.dev/how-does-warp-compare/performance
Though I don't care about it, but looks like performance isnot the pros of wezterm.

Seems Alacrity has the best performance just now, and some feaurs are similar to wezterm, but it's not stable enough, even I cannot get it work on my windows 11 and mac.
I'm indeed like the vi like key binding and lua configuration. wezterm is great.

**Reply by aleksandersumowski** [reply]
Alacrity doesn't implement terminal multiplexing though, so for a lot of people a fairer comparison would be WezTerm vs Alacrity + tmux

**Reply by qiuwch** [reply]
Alacrity can not support showing image in the terminal, as shown here, https://yazi-rs.github.io/docs/image-preview. This is a dealbreaker for me.

**Reply by Kreijstal** [reply]
you have to use the alacritty sixel fork

**Comment by robrecord** [comment]
It's pretty hard to get started, especially to make it look good - that was my experience. 

Also the promise of multiplexing isn't quite delivered. I'm still trying to figure out how to fit my tmux workflow into WezTerm (session persistence, having more than one session open at once...)

**Reply by ema2159** [reply]
As per the customization, it is quite good and easy. You cannot compare it with other "out of the box" terminal emulators, but rather other highly customizable terminal emulators (Alacritty, Kitty). Lua is a fantastic language for configuration and extension, and compared to Kitty and Alacritty, I found Wezterm extremely easy to configure.

It also comes with a sensible font default (JetBrains Mono) which makes it look nice out of the box too.

**Reply by SuperKenVery** [reply]
I found the multiplexing very easy to use... I wonder what problem you're encountering

**Comment by EarlGray** [comment]
Memory consumption: on my machine right now, it's ~170 MB of resident memory vs ~80 MB for Alacritty.

Some time ago I could feel noticeable delay for events. I suspected this was due to the smol executor or event handlers in Lua. This is not a real issue anymore.

Customizability is great, but it can have downsides in resource consumption and response latency if left unchecked.

Also, it felt weirdly out of place on Mac. I could not quite put my finger on what exactly was wrong, but it did not feel integrated at all (non-mac font rendering? non-mac keyboard shortcuts? drag-and-drop not working as expected?), so much that I had to switch back to Terminal.

**Reply by eugenesvk** [reply]
Is higher memory use because of customizability??

**Reply by kalocsaibotond** [reply]
Yes, memory and resource consumption is the greatest weakness of WezTerm. It is even worse with WebGpu frontend. With that, WezTerm reaches a whopping ~320 MB of resident memory (while Alacritty for me uses a ~50 MB consistently out of box).

I don't think that this resource requirement comes from customisability, since lua is a tiny language. Lua gives the amazing configurability of WezTerm (and its lua api). The lua interpreter ought not to occupy more than 1 MB memory.

It would be nice to have configuration options that strips down all the fancy overhead, decrease resource use, and augments performance to a ConEmu level.
Considering, the fact that how surprisingly popular the suckless simple terminal is, there are many people who would be satisfied with a minimalistic, but performant terminal emulator that can render nerd fonts quickly.

**Comment by AndydeCleyre** [comment]
I love wezterm. Here's what I don't like:

- The configuration being in a full programming language (lua), rather than a configuration language (e.g YAML, TOML, JSON5, HJSON, NestedText, UCL). This also hinders querying or modifying the configuration from other code.
- All the default shortcuts that I need to explicitly disable in the config. I keep adding as I find interference problems, but so far:

```lua
    keys = {
      {key="Enter",      mods="ALT",        action="DisableDefaultAssignment"},
      {key="f",          mods="CTRL",       action="DisableDefaultAssignment"},
      {key="f",          mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="k",          mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="LeftArrow",  mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="RightArrow", mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="UpArrow",    mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="DownArrow",  mods="CTRL|SHIFT", action="DisableDefaultAssignment"},
      {key="PageUp",     mods="SHIFT",      action="DisableDefaultAssignment"},
      {key="PageDown",   mods="SHIFT",      action="DisableDefaultAssignment"},
    },
```

But really it's fantastic.

**Reply by bew** [reply]
An alternative to interference problems is to disable all default key bindings and define only the ones you want, that's what I do for keys and mouse bindings `¯\_(ツ)_/¯`

**Reply by gegoune** [reply]
That is correct approach indeed. There is a lot of default key bindings by default, I have removed them all and configured only those is need.

As in for lua - this is one of the main advantages and what makes Wezterm stand out, in my opinion.

**Reply by eugenesvk** [reply]
> The configuration being in a full programming language (lua

But that's one of the best parts of Wezterm (except for lua, which isn't a good language)!

**Reply by bew** [reply]
@eugenesvk I think Lua is pretty great for this config use case, what do you not like about it?

**Reply by eugenesvk** [reply]
What's great about it besides "easy to embed"?

It has awful standard library: I have to google around for, copy&paste string manipulations basics like `startswith`, then find out that they didn't escape magic symbols properly, so have to write my own `escmagic` string extension. Or the other bare minimums like rounding a number to .xy, or working with paths, or ... anytning else: it's just a **waste** of effort having to **recreate** your own **std**

Then the **data type** support is bad with "everything is a tables", but then see above due to empty std it doesn't even have many utils for working with tables, so yet again you need to find some. (case in point: I think I used your flatten/merge all/deepclone library if I recall correctly to have more ergonomic config editing experience where you can split tables into semantic blocks and combine them later)

Then there are some **table quirks** like table length being unreliable and only working on contiguous seqences (or something)
Or the global-by-default vars

And then since lua isn't stable, the results you find are often not working for whatever the version WT is using

Then the package ecosystem is also lackluster and outdated, so no much help there

And of course you wouldn't remember much of that when you get back to your config in a year to add another bright idea

So while I'd take lua over JSON any day of the week, it's sad that it's often the first and only choice for apps adding programmable config

**Reply by AndydeCleyre** [reply]
@bew Yes, thank you, I've now cut down my keybinds to

```lua
  disable_default_key_bindings = true,
  keys = {
    { key = '+', mods = 'CTRL', action = wezterm.action.IncreaseFontSize },
    { key = '-', mods = 'CTRL', action = wezterm.action.DecreaseFontSize },
    { key = '0', mods = 'CTRL', action = wezterm.action.ResetFontSize },
    { key = 'v', mods = 'SHIFT|CTRL', action = wezterm.action.PasteFrom 'Clipboard' },
  },
```

I think there's a stringy way to specify those actions so I don't have to have `wezterm.action.` prefixes but I couldn't figure it out in the case of `PasteFrom 'Clipboard'`.

**Reply by AndydeCleyre** [reply]
Regarding Lua as a config format -- I have an external config file (yaml, json, nestedtext, whatever) which keeps ordered lists of my preferred fonts for each type (mono, sans, serif).

I then script value injection or template rendering to use this single source of truth to update any relevant config files (e.g. `~/.config/fontconfig/fonts.conf`). Is it possible to do the same with `wezterm.lua`? That is, take values from a data list or code, re-fill the relevant bit of lua in `wezterm.lua`, preserving all other content, including comments?

**Reply by wez** [reply]
you can read files in lua and parse the contents.
There is a builtin json parser: https://wezfurlong.org/wezterm/config/lua/wezterm/json_parse.html
There is a PR to add yaml and toml: https://github.com/wez/wezterm/pull/4969

**Reply by AndydeCleyre** [reply]
> you can read files in lua and parse the contents.
> There is a builtin json parser: https://wezfurlong.org/wezterm/config/lua/wezterm/json_parse.html
> There is a PR to add yaml and toml: https://github.com/wez/wezterm/pull/4969

Thanks! My truth sources are not those formats, but this does open the door for me to generate such a file, then import that.

**Reply by wez** [reply]
lua also has built-in file io functions that you can use to implement any file format if you prefer. You can google and find a lot of examples of how to read files in lua

**Reply by AndydeCleyre** [reply]
I'll note that this still is not ideal, as I'd like to be able to produce a single wezterm config file that can be copied around in a complete form, without relying on any external files or programs.

So I could turn my wezterm "config" file (which is really a program) into a template file, only edit that directly, and generate the `wezterm.lua` as output.

Another downside of suggested alternatives is of course that not everyone who wants to configure Wezterm also wants to learn to develop Lua programs.

**Reply by AndydeCleyre** [reply]
Well if anyone else struggles to manage the verbosity and complexity of writing lua to reconfigure wezterm, I've made myself a [template](https://github.com/AndydeCleyre/dotfiles/blob/master/.config/wezterm/wezterm.lua.wz) for [wezterm.lua](https://github.com/AndydeCleyre/dotfiles/blob/master/.config/wezterm/wezterm.lua) and now [regenerate](https://github.com/AndydeCleyre/dotfiles/blob/master/.config/wezterm/make.sh) it after editing a [vars.nt](https://github.com/AndydeCleyre/dotfiles/blob/master/.config/wezterm/vars.nt) config file. So far, so good.

**Reply by JoyceBabu** [reply]
@AndydeCleyre I think you are doing it wrong. Creating a JSON/YAML file and reading it from wezterm.lua might be easier and less error prone than generating the wezterm.lua from YAML.  You can bypass the transpilation step completely, making it much more portable.

Lua is not a difficult language to learn, and is widely used as an embedded programming language. Learning it will come in handy later.

**Reply by AndydeCleyre** [reply]
@JoyceBabu 

I hear you, but among the reasons that's not the path for me is this:

> I'll note that this still is not ideal, as I'd like to be able to produce a single wezterm config file that can be copied around in a complete form, without relying on any external files or programs.

**Comment by eugenesvk** [comment]
I'm not sure what the biggest limitations to getting more attention are (I also think it's the best terminal out there), so will just list a few of areas for improvement:

- lua is too simplistic of a language for anything but simplistic customizations, and the ability for more complex customization is what makes WezTerm great! Ideally WezTerm would have some WASM plugin host instead so you could use **any language** with a wide **ecosystem** available.
  - For example, since I dislike the unreadable hex color syntax codes for setting terminal color theme, I wanted to be able to use something nice like OKHSL, but then there was no lua package, so had to port
  - It's especially bad when you want to tweak a thing or two in your config after a while and have to relearn all the lua language design issues again
  
- more ergonomic keybindings customizations would be nice as it's easy to get lost in the layers (though stacked approach is great in itself)
 
- memory use could maybe be lower (though I'm not sure I'd trade any functionality to achieve that :) )

**Reply by wez** [reply]
FWIW, I looked into WASM and haven't found a runtime that supports async execution.  I'm not sure WASM even supports that concept. That's important in an event based gui program where it is critical to avoid blocking the gui thread.  This is an area where lua really shines.

re: lackluster std library, if you can identify the set of functions you've sworn at the most and had to implement yourself, I'd be happy to have wezterm export the equivalents from the rust std library.  Please file an issue with that list!

re: colors, doesn't https://wezfurlong.org/wezterm/config/lua/wezterm.color/parse.html support what you need?  You should be able to use most CSS color syntaxes

**Reply by eugenesvk** [reply]
Thanks for looking into WASM! Ok, so then this is still a pie-in-the-sky type of idea and maybe there isn't a good way to be more polyglot.

Or maybe instead of a runtime there could be a compile/transpile type of support where you could write your config in python or Rust or JS and then have it transpiled to lua which WezTerm will use to maintain async (and I fully agree it's crucial for a GUI app).

I'm not sure since I haven't seen apps with such awesomly flexible configurability (and it might be too complicated, I'm already using a preprocesor step with a config manager that allows me to automatically adjust terminal font size for desktop1/desktop2/laptop monitors, and it adds friction)

**Reply by eugenesvk** [reply]
> You should be able to use most CSS color syntaxes

None of them are good :) afaik CSS v4 only supports oklch, but not okHSL/V https://github.com/w3c/csswg-drafts/issues/8659

I've checked the methods on https://wezfurlong.org/wezterm/config/lua/color/index.html#available-methods and there doesn't seem to be okhsl/v either

Anyway, the syntax I use, which allows me to use the minimum pure vertically aligned numbers and to:

- set a global default color space
- have per-color overrides (e.g., `okhsv  29 100 100` instead of okhsl to emphasize the max values)
- optional niceties like `°` indicators that get stripped on parsing (CSS supports the old-school `deg`, but they'll likely never get to this kind of visual pleasantries)

```lua
local colSpaceDef	= "okhsl"	-- treat all (non-prefixed) triplet numbers as HSL
local colSpaces  	= {"hex","hsl","hsv","hwb","okhsl","okhsv"} -- allowed color spaces (prepend with ␠ or :)

local col         	= {}
col.background    	= "        0   0   5"	-- okHSL due to colSpaceDef
col.cursor_bg     	= "      227 100  74"	-- 
col.visual_bell   	= "okhsv  29 100 100"	-- hsV to emphasize the 100 (okHSL = 29 100 57)
col.compose_cursor	= "hsl    38 100  50"	-- copy paste from another scheme, haven't converted
```

(we've discussed the lack of global prefixes in this issue https://github.com/wez/wezterm/issues/1436 that links to some older version of my lua converter)

**Reply by celgost** [reply]
"lua is too simplistic of a language for anything but simplistic customizations"

I don't think this statement is quite fair. Lua is often used in game dev for its simplicity in extending a game. Every game in Roblox is made in Lua and some are actually pretty neat (check out FRONTLINES). Every Interface addon in WoW is made in Lua, and you have some complete GUI overhauls. On a software closer to wezterm, AwesomeWM config is in Lua, and the amount of things you can do on this window manager is quite crazy. 

I understand you might not like the language. It has its limitations, and if you don't like the logic of "everything is a table" well... you're doomed. But it is a very powerful language to make things with really!

**Reply by eugenesvk** [reply]
The fact that powerful things can be built doesn't negate the argument that the language doesn't help you do that, that was the point, you encounter too many issues and limitations of it simplistic design when you try to build something more powerful

**Reply by celgost** [reply]
Don't get me wrong I agree with this!

The point behind my argument was also more "Is it really that much of a dealbreaker for a terminal emulator configuration?". It seems a bit overkill to me in that particular use case. 99% of people will probably be just fine with Lua on Wezterm. It's easy to understand as well as easy to share. And usually, you're not fiddling with your config every morning. You put a bit of work into it, find something that's good to you and then you just use the software.

I'd also mention that even if by example you can do awesome things in Roblox, game devs on Roblox Studio are absolute masochists in the first place lol. (not to mention the predatory 75% revenue cut but that's something else)

**Reply by eugenesvk** [reply]
(yes, it's not a dealbreaker)

> You put a bit of work into it, find something that's good to you and then you just use the software.

Not really, you don't have the perfect design at the start, what usually happens is what I've mentioned in the original comment:
> It's especially bad when you want to tweak a thing or two in your config after a while and have to relearn all the lua language design issues again

But even at this initial stage, the less ergonomic the language is, the less likely it is that the marginal user will use it and implement (and then share) his customization improvements, so that means we are less likely to benefit from that

**Comment by AndydeCleyre** [comment]
I'll just add that Wezterm has so many cool abilities . . . that I use tmux for. So I worry about the complexity becoming a maintenance burden. But you know, so far so good. Wez continues to be excellent at both the development and open discussion for this project. 🙏🏼

**Reply by NonlinearFruit** [reply]
I agree but also someday supporting Control Mode would be awesome (#336)

**Comment by lost22git** [comment]
Probably because the stability under windows is not as good as windows terminal, wzt can easily crash the shell ;-)

**Reply by Destroy666x** [reply]
I have 0 clue where that's coming from... To me this take is pure nonsense. WezTerm on Windows - 0 issues. Well, one small issue - it closes when graphic drivers get updated, but that's nothing. Other than that, I never encountered a serious bug. While with Windows Terminal I get bugs every 5th action or so, e.g. it often randomly spawns 2 windows instead of 1 - that terminal is extremely overrated and overpromoted. Even outdated ConEmu/Cmder perform better, IMO.

WezTerm is the closest Linux experience on Windows, comparable to tmux and other commonly used terminal apps. If it got the missing feature of session management and maybe some more scritable events, it'd be perfect for me.

**Reply by kilbyjmichael** [reply]
I have quite a few issues on some windows machines but not all. My work laptop shows graphical bugs and crashes wezterm daily but at home it has less issues with that. On linux it does not seem to have any of these issues.

**Reply by sebastianstudniczek** [reply]
@Destroy666x Weird, I have 0 problems with Windows Terminal, and I am using it for almost 1.5 year. Really stable and fast choice for windows. 
Recently I tried WezTerm but have some weird issues, for example when moving from one monitor to another the window it's stuttering or when using PSFzf there are some rendering issues (but maybe that's problem with PSfzf itself).

**Comment by aj3423** [comment]
Currently, to copy all text:
1. enter copy mode
2. `g` -> `V` -> `G` -> `y`

That's a lot of key strokes, in my workflow I need to do this a lot. It would be nice if there is a built-in "copy all" feature.

**Reply by bew** [reply]
If you want to copy everything in a single keybind you can probably do it yourself using:
- https://wezfurlong.org/wezterm/config/lua/wezterm/action_callback.html
- https://wezfurlong.org/wezterm/config/lua/pane/get_lines_as_text.html
- https://wezfurlong.org/wezterm/config/lua/window/copy_to_clipboard.html

**Reply by aj3423** [reply]
Thanks. Now I understand why it uses lua as configuration file, which indeed makes it the best, this is my script for "copy all":
```lua
		{
			key = 'c',
			mods = 'CTRL|SHIFT',
			action = wezterm.action_callback(function(win, pane)
				local lines = pane:get_lines_as_text(1000000) -- same as `scrollback_lines`
				win:copy_to_clipboard(lines, 'Clipboard')
			end),
		},

```

**Reply by eugenesvk** [reply]
I don't think you need to have constats that match the scrollback, you can also get that info from the [get_dimensions](https://wezfurlong.org/wezterm/config/lua/pane/get_dimensions.html)
```lua
local function copyAll(win,pane)  -- copy all text in a pane (except for bottom prompt in Xonsh)
  local prompt_bottom_offset = 0
  local proc = pane:get_foreground_process_info()
  local name, binpath, arg = proc.name,proc.executable,proc.argv
  local isXonsh = sh.isXonsh(arg)
  if isXonsh then
    if os.getenv("BOTTOM_TOOLBAR") then
      prompt_bottom_offset = 2 -- exclude the main + bottom prompts
    else
      prompt_bottom_offset = 1 -- exclude the main          prompt
  end end

  local dims = pane:get_dimensions()
  local txt  = pane:get_text_from_region(0
    , dims.scrollback_top              , 0
    , dims.scrollback_top + dims.scrollback_rows - prompt_bottom_offset)
  win:copy_to_clipboard(txt:match('^%s*(.-)%s*$')) -- trim leading and trailing whitespace
end
```

**Comment by AndrewSav** [comment]
Agree with the following points raised above:

- lua is neither here not there:
  - on one hand the configuration language is difficult to script, yaml/json/toml etc would be preferable
  - on the other hand as a programming language lua is severely limited, wasm would be preferable here because you can compile almost anything into wasm
- This is a programmer's terminal, not a end user's terminal. If you are not prepared to invest hours in configuring it (and enjoy the process), it is probably not for you. In particular:
  - Default key bindings do not make much sense out-of-the-box
  - I doubt that unconfigured configuration defaults are suitable for many people, they were not for me
  - Configuring requires usage of a programming language and understand programming model (event, objects, etc)
  - Because of that, sometimes just copied two (seemingly unrelated) configuration pieces from the documentation to your config can lead to unexpected results. Programmers deal with this every day, a regular user, very likely won't be able to make do without assistance

I'll add one from myself: [wezterm cannot do broadcasting and it is not going to be added](https://github.com/wez/wezterm/issues/2658). If you need broadcasting, it's a total dealbreaker, you just don't use wezterm and use a terminal that can.

I think that the original question trying to address why wezterm is not more popular. For me the answer is clear: it's a brilliant tool, but you need to spend time to learn and maintain it, something that most people would not be willing to. Therefore it is _niche_,

I doubt very much that wezterm becoming very popular tool is a goal for the author, so because of this it will probably never be. If it was a goal, improvements could be made:

- Lua has to go. Seriously. One does not have to learn the internal eventing system and object model to configure a god damn terminal
- Documentation should be written with non-programmers in mind. Currently it does not even explain what _multiplexing_ is, or why would we need domains, it's assuming, that it's obvious
- Work needs to be done on sane defaults that are usable out of the box
- UI / Features need to be more discoverable. (E.g the program needs a way to guide you to Ctrl-Shift-P, similar to how Windows Terminal does that)
- We need a primer that can help answer questions like "I currently do this in tmux, how can I do this same things in wezterm?"

**Reply by Destroy666x** [reply]
You are not wrong most of the time in your comment, but:
- "Lua has to go" is nonsense. It doesn't have to go to simplify anything... The entire purpose of Lua is to provide scriptability as an addition to basic functionality. Optional scriptability. You can just add some missing features that have to be scripted now and move configs to a more config-appriopriate format, such as YAML. That's it. You really don't have to burn the whole forest down to eliminate a wolves eating your chickens kind of problem. A beyond terrible solution that alone is good reason to downvote this.
- Defaults are sane 95+% of time. Examples of bad ones would be welcome.
- You mention that multiplexing needs to be explained to newer users, yet in the same list you include explaining stuff to tmux users, which are the opposite. Tmux is I'd say harder to master than WezTerm, which is not suprising considering it's coming from Linux poweruser background. So unless you want 2 documentations or just a subpage like "Hints for tmux users", this isn't really doable without a mess IMO.

**Reply by pkazmier** [reply]
For me, Lua is a huge selling point. I'm able to do cool stuff like #5435.

**Reply by krskrs** [reply]
+1 in favor of Lua. It was the feature that made me want to try Wezterm

**Comment by LeoGitHuber** [comment]
My Wayland compositor is Hyprland latest version. Wezterm on my system compared to Alacritty is noticeably less performant. When I'm typing, I can feel kind of lag. So I decided to keep using Alacritty.

**Reply by xymeng16** [reply]
Do you enable webgpu? I was feeling laggy in WezTerm before enabling it but after that it has been working smoothly.

**Reply by Ilirski** [reply]
@xymeng16 Might be a Hyprland thing. For me, even with WebGpu enabled (and running on Wayland), there is noticeable lag when terminal scrolls when searching for text on Neovim.

**Reply by sgon00** [reply]
Yeah, I am running KDE(X11). I can feel lag too. Your comment finally makes me switch back to Alacritty.

**Comment by mishurov** [comment]
My Wayland compositor is Sway Debian Stable version. I wait for right-click context menu and selecting items with mouse, without that feature Foot can do everything I need and is in the official repository. So I decided to keep using Foot.

**Comment by quadespresso** [comment]
I'm having another go trying to get back into Wezterm. I'm presently using Terminator, and had previously used Tilix. I would probably still be using Tilix, save for one annoying bug: URL's beyond a certain length (FQDN) become unclickable.

Several months ago, I started down the path of trying to customise Wezterm; I managed to somehow b0rk it in such a way that required me to use another terminal program to revert the config change. I didn't think I had done anything particularly wicked, as I was trying to ease into things.

 I came to realise that any modest customisation would result in a side quest to learn. Conversely, any extended period of satisfaction would lead to losing any familiarity I once had.

Fwiw, I have extensive use with iTerm on the Mac. Although one could make the argument that the config isn't as powerful as Wezterm due to the use of Lua in the latter, I've managed to customise my iTerm setup in rather pleasing ways. I don't recall whether Wezterm can do this, but I had configured iTerm to restart/recover (eg, after a reboot) from the last known window/tab/pane layout.

I say that I'm having another go getting into Wezterm. But the truth of the matter is, I've got things to do. Any digging that starts to look like it's turning into a side quest will discourage me from digging further. Perhaps the answer for me personally is to just leave it be and focus on expanding familiarity, customisation of tmux.

**Comment by quadespresso** [comment]
When I was dabbling with Wezterm, I discovered that a misconfigured config file would force me to use some other terminal to recover.

I can't speak for all others, but Alacritty simply complains about the misconfig, referencing the approximate place in the file where the misconfig was found. Granted, it's possible I just haven't managed to craft a sufficiently janky config file for it to stumble over.

**Reply by axelhj** [reply]
That's interesting. I found that the config auto-reload and error messages was particularly good with Wezterm (compared with configuring Neovim).

But I didn't try reopening with a borked config file. I did need to open Wezterm in Wezterm (or any terminal) to see errors in custom event handlers though, in order to debug my Lua scripts (maybe this goes for any runtime errors/non startup errors). Still, I'm quite pleased with Wezterm overall. I will need to try out Alacritty at some point to compare performance though.

**Reply by Risinggoblin** [reply]
Yes, a nix-style rollback will be a dream come true for wezterm configs!

**Comment by git001** [comment]
As I have used several terminal emulation like wez.
I have to switch to wez because of that issue https://bugs.kde.org/show_bug.cgi?id=482628 in konsole.

wez looks nice for now but a "simple task" like "set geometry height X width (170 x 40)" is very difficult because it's not clear to me is it window/pane/frame/screen/???

I have seen https://wezfurlong.org/wezterm/config/lua/window/set_position.html but this is for position in the Desktop.

Maybe I can tell wez to save the current size when I close the terminal?

Maybe there is a snipplet collection, which I haven't seen yet, which shows how to configure such simple parts in wez?

**Reply by git001** [reply]
I have found the solution for my question.

Here the config snipplet which works for me.

```
-- Pull in the wezterm API
local wezterm = require 'wezterm'

-- This will hold the configuration.
local config = {}
if wezterm.config_builder then
  config = wezterm.config_builder()
end

config.initial_rows=25
config.initial_cols=170

-- and finally, return the configuration to wezterm
return config
```

**Reply by akthe-at** [reply]
Here is a snippet I have in my config that sets the window position and the size of the screen when I launch wezterm. I also included a snippet that is commented out to launch it maximized...

```
local wezterm = require "wezterm"
local mux = wezterm.mux
local config = {}

wezterm.on("gui-startup", function(cmd)
  local screen = wezterm.gui.screens().active
  local tab, pane, window = mux.spawn_window(cmd or {})
  -- window:gui_window():maximize() -- maximizes the window at startup but I don't currently use this.
  local gui = window:gui_window()
  local width = 0.95 * screen.width
  local height = 0.90 * screen.height
  gui:set_inner_size(width, height)
  gui:set_position((screen.width - width) / 2, (screen.height - height) / 2)
end)

return config
```

**Comment by btegs** [comment]
When I am using LazyVim with Neovim's Catppuccin color scheme, Alacritty seems to render the colorscheme a little nicer. Colors seem to match better and the font rendering is smoother.

I'm a big fan of using Pop!_OS on my Linux machine and the upcoming COSMIC desktop will ship with cosmic-term which uses Alacritty as the backend: https://github.com/pop-os/cosmic-term

**Comment by Emille1723** [comment]
> As I have used several terminal emulation like wez.
> I have to switch to wez because of that issue https://bugs.kde.org/show_bug.cgi?id=482628 in konsole.
> 
> wez looks nice for now but a "simple task" like "set geometry height X width (170 x 40)" is very difficult because it's not clear to me is it window/pane/frame/screen/???
> 
> I have seen https://wezfurlong.org/wezterm/config/lua/window/set_position.html but this is for position in the Desktop.
> 
> Maybe I can tell wez to save the current size when I close the terminal?
> 
> Maybe there is a snipplet collection, which I haven't seen yet, which shows how to configure such simple parts in wez?

I switched to wez for the same bug in Konsole. I'm using Wezterm on my Linux machines ATM to get around that Konsole bug but funnily enough, the bug isn't present on Konsole for WSL. I'm happy to use Konsole still in WSL because the performance of Konsole in WSL (installed and run through the Linux OS) is unmatched, something that I spent a day trying to optimize in Wezterm and unfortunately I failed. I'd love to hear of anyone has any tips. I'm not allowed to run Wezterm natively on Windows (work) so getting it performant running directly inside WSL like Konsole is my only option.

On my Linux machines, Wezterm has been great though there is a slightly noticeable delay compared to Konsole which is really fast.

I decided also to move to Wezterm cause it handles somethings better than Konsole like image rendering which I noticed recently was a problem in Konsole when using yazi

**Comment by SuperKenVery** [comment]
I like wezterm a lot, but I can put some pain points here:
 - No tmux support (sshmux+nix-portable work quite well, but tmux is almost preinstalled everywhere)
 - Bad window management on macOS: sshmux creates a new app icon in dock
 - Bad resize experience: texts fly around when resizing, and sometimes the terminal fail to resize with the window
 - You can't drag around tabs; you can only rearrange them using commands

**Comment by osenvosem** [comment]
For me Wezterm turned out to be less performant than Kitty: it consumes more ram and feels slower, even had lags from time to time.

**Comment by tjex** [comment]
The only thing that makes it 'bad' would also be what makes it good, and that's the extensive customisability with lua.
It's easy to get carried away... Sure it's a bit more heavy than other terminals, but with the computers we're working on these days, it's a drop in the RAM ocean. Despite it being a bit heavier, I can't say I notice any lag (I'm on a m1 max, 64gb RAM).

I do worry sometimes for the future of WezTerm though, that because of all the extra customisability and multiplexer integration, that it will become too much for Wez to handle. If WezTerm is an essential part of people's setup (e.g, neovim users, sys admins, etc) please consider supporting Wez. Would be SUCH a sting if he had to drop the project...

**Comment by syhol** [comment]
I've been daily driving for over 6 months now, and I only have a few small complaints:

### Search

- https://github.com/wez/wezterm/issues/1988 - When opening a search again and start typing, I don't expect it to start appending to my last search term. I know I can ctrl+u but I'm struggling against muscle memory.
- https://github.com/wez/wezterm/issues/3684 - It's really painful searching through streaming logs

### Pane Padding

- https://github.com/wez/wezterm/discussions/1383 - I came from kitty which has pane padding and I loved it so much more

---

Still, Wezterm is currently my favourite terminal with its multiplexing, decent font rendering, dpi_by_screen, and lua config.

**Comment by Kreijstal** [comment]
no sixel on windows (when even windows terminal has sixel support)

**Comment by quadespresso** [comment]
I'm back to using Wezterm, have done for at least a few months now.

I'm using it mainly as a fast, lightweight terminal. I've got a minimalist config to set the title bar and use the Catppuccin theme.

I have, for the time being, settled on uzing Zellij for my muxing needs. I'm mainly looking for a satisfying combo of terminal + mux + Neovim. I'm considering tmux (with plugins), as that plus Vim/Neovim seem to be pretty popular for workflows. I don't want to get too tied to any one terminal's mux approach/config; I would probably lean into a terminal that had a more friendly embrace of tmux/Zellij while still supporting inline graphics (yazi). 

Once Ghostty is GA, I'll have a look at that as well. I've already seen feedback suggesting that there's nothing particularly groundbreaking there. I'm less interested in having a terminal that doubles as a Swiss Army chainsaw, and more interested in a terminal that does a core set of things well, with as modest a system footprint as is reasonable.

Quick aside: I just now had another look at the Foot terminal. It seems to support inline graphics, something I could have sworn I'd checked previously. Foot + tmux + yazi works well for graphics, but Foot + Zellij + yazi triggers some peculiar scroll/resizing when viewing JPG files of differing dimensions. Again, nothing so off-putting that I would avoid Foot altogether, but it might steer me more towards using tmux sooner rather than later.

**Comment by sgon00** [comment]
HiDPI issue: https://github.com/wez/wezterm/issues/5453, I can NOT use it. I just installed wezterm today. If I can not fix this issue in next 2 hours, I have to give up.

### Updated:

In order to use WezTerm, I had to switch to KDE (X11).

My original problem was that XFCE4's scaling setup only works with GTK apps, so I had to use `config.dpi`. However, due to the above GitHub resize issue, I am unable to use `config.dpi`.

After switching to KDE (X11), things changed. Although the resize issue with `config.dpi` still exists in KDE (X11), WezTerm now honors KDE's system scaling setup, so I can use it on my 4K monitor.

**Reply by SuperKenVery** [reply]
Use wayland

**Reply by sgon00** [reply]
Thanks for reply. But wayland is too slow. At least for ubuntu gnome. Many apps do NOT support it well.

**Reply by SuperKenVery** [reply]
You could say your apps aren't supported well, but I don't know why you say it's slow: it's a better optimized structure than X11.

If you're referring to HiDPI support for XWayland, then you should try KDE, it has built-in fix for fraction scaling under XWayland.

**Reply by sgon00** [reply]
The reason why I said ubuntu gnome wayland is slow because I feel it's slow. Everything is slow. Mouse movement, mouse click, popup window etc.. It's like slow motion. I can feel the pause. I am not sure how long the pause is. Maybe 0.5s or 1s. Everything is installed well and drivers are OK. It's just slow. Sorry that I am not a native English speaker and I am not sure if I express it clear. I am not running on old hardware. AMD Ryzen 7 6800H and 32G RAM. I have no issues in XFCE.

**Reply by SuperKenVery** [reply]
I have no idea why that's happening... sorry to hear that. 

I think you expressdd it super clear, though I'm not a native English speaker either :D

**Reply by sgon00** [reply]
@SuperKenVery Thanks for your reply. In order to use WezTerm, I have switched to KDE(X11) (install `kubuntu-desktop` package in ubuntu 24.04).  I have updated the original comment post for details. Now, everything works. :)

**Comment by sgon00** [comment]
Another major issue: IME not working: [3221](https://github.com/wez/wezterm/issues/3221) and [5125](https://github.com/wez/wezterm/issues/5125).

**Reply by gegoune** [reply]
Please stop posting every issue you run into in this discussion. It's not what it's for.

**Reply by sgon00** [reply]
Thanks for your comment, but the discussion title is `What makes wezterm bad?`. So what is this discussion for?

**Reply by gegoune** [reply]
For conceptual limitations of the terminal emulator. Not to keep track of who ran into which issue.

**Reply by sgon00** [reply]
1. The topic does not mention the term **conceptual** limitations, nor do most replies.

2. I am not posting all issues, just the showstoppers that "make WezTerm bad," e.g., the IME not working (preventing typing) and the HiDPI issue (making it impossible to read). A terminal should at least allow the user to input and read words.

3. I am leaving WezTerm. In any case, this should be my last comment regarding WezTerm, at least for 2025. Apologies for any trouble I may have caused.

**Comment by wez** [comment]
I'm going to close and lock out this particular discussion; every time I got notified by it, I get a sense of dread because of the title. It's demoralizing and sucks away any energy I might have to work on wezterm.

I appreciate the spirit of the original post, but the words are bad and make me feel bad every time I see it.

I'm in favor of *constructive* criticism: "what would make make X better is Y", rather than "X sucks because Y".

You're welcome to open new discussions with that in mind, but for now: thanks folks for participating here.
